File: 백준/Silver/10815. 숫자 카드/숫자 카드.py
import sys
sys.setrecursionlimit(10**7)


def main():
    N = int(input())
    arr = list(map(int, sys.stdin.readline().split()))

    M = int(input())
    card = list(map(int, sys.stdin.readline().split()))
    res = []

    arr.sort()

    for i in card:
        res.append(find(arr, 0, N-1, i))

    res_str = list(map(str, res))
    print(' '.join(res_str))


def find(arr, start, last, num):
    if start > last:
        return 0
    else:
        mid = (start + last) // 2
        if arr[mid] < num:
            return find(arr, mid + 1, last, num)
        elif arr[mid] > num:
            return find(arr, start, mid - 1, num)
        else:
            return 1


if __name__ == "__main__":
    main()

File: 백준/Silver/10815. 숫자 카드/숫자 카드.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        ArrayList<Integer> arr = new ArrayList<>();

        StringTokenizer st1 = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr.add(Integer.parseInt(st1.nextToken()));
        }

        int M = Integer.parseInt(br.readLine());
        ArrayList<Integer> card = new ArrayList<>();

        StringTokenizer st2 = new StringTokenizer(br.readLine());
        for (int i = 0; i < M; i++) {
            card.add(Integer.parseInt(st2.nextToken()));
        }

        arr.sort(Comparator.naturalOrder());
        ArrayList<Integer> res = new ArrayList<>();

        for (int i = 0; i < M; i++) {
            res.add(find(arr, 0, N - 1, card.get(i)));
        }

        for (int i = 0; i < M - 1; i++) {
            System.out.print(res.get(i) + " ");
        }
        System.out.print(res.get(M - 1));

    }

    public static int find(ArrayList<Integer> arr, int start, int last, int num) {
        if (start > last) {
            return 0;
        } else {
            int mid = (start + last) / 2;
            if (arr.get(mid) < num) {
                return find(arr, (mid + 1), last, num);
            } else if (arr.get(mid) > num) {
                return find(arr, start, mid - 1, num);
            }
            return 1;
        }
    }
}
